autokeras/hypermodel/base.py

```diff
@@ -1,13 +1,95 @@
+import pickle
+
 import kerastuner
 import numpy as np
 import pandas as pd
 import tensorflow as tf
+from kerastuner.engine import stateful
 from tensorflow.python.util import nest
 
 from autokeras import utils
 
 
-class Node(kerastuner.engine.stateful.Stateful):
+class Pickable(stateful.Stateful):
+    """The mixin for saving and loading config and weights for HyperModels.
+
+    We define weights for any hypermodel as something that can only be know after
+    seeing the data. The rest of the states are configs.
+    """
+
+    def get_state(self):
+        """Returns the current state of this object.
+
+        # Returns
+            Dictionary.
+        """
+        return {'config': self.get_config(),
+                'weights': self.get_weights()}
+
+    def set_state(self, state):
+        """Sets the current state of this object.
+
+        # Arguments
+            state: Dict. The state to restore for this object.
+        """
+        self.set_config(state['config'])
+        self.set_weights(state['weights'])
+
+    def get_config(self):
+        """Returns the current config of this object.
+
+        # Returns
+            Dictionary.
+        """
+        raise NotImplementedError
+
+    def set_config(self, config):
```

Comment:
This should probably follow the Keras convention:

```python
@classmethod
def from_config(self, config):
  ...
```

"State" and "config" should probably be entirely separate (we don't make this distinction very well in Keras Tuner at the moment, but AutoKeras can do better I think)

The methods should probably be:

```get_config``` -> Returns the arguments the object was constructed with
```from_config``` -> Makes the object
```get_state``` -> Returns non-config related state. I.e. only things that changed from instantiation. This is the equivalent `keras.layers.Layer.get_weights()`
```set_state`` -> Sets the non-config relate state. This is the equivalent of `keras.layers.Layer.set_weights()`

This way you can reuse the Keras mechanisms for `get_config`, `from_config`, and `custom_objects` in order to support user-provided objects

`save` and `reload` could make use of the other methods, but `get_weights` and `set_weights` would be redundant